### PR TITLE
Canvas Typesetter

### PR DESCRIPTION
**Pull request by themadcreator**
Add support for Canvas elements

In recognition of this more generic functionality and significant API changes, rename library to `@palantir/typesetter`

- Remove d3 dependency!!!
- Create interface for measuring (`IRuler`) and writing (`IPen`) functionality.
- Implement those interfaces for SVG and Canvas.
- Restructure API to accept these so-called `ITypesetterContext`s.
- Add new `Typesetter` top level object to simplify the creation of the default `CacheMeasurer` / `Wrapper` / `Wrapper` stack in both SVG and Canvas contexts.
- Add unit tests to keep coverage > 90%.
- Regenerate typedoc

### CODE REVIEW

**Comment by themadcreator** [comment]
New readme: https://github.com/palantir/svg-typewriter/blob/9cbc2b1303604c6a882df771ea663bfda88d470e/README.md

File: src/typesetter.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * license at https://github.com/palantir/svg-typewriter/blob/develop/LICENSE
+ */
+
+import { CanvasContext, ICanvasFontStyle, ITypesetterContext, SvgContext } from "./context";
+import { CacheMeasurer } from "./measurers";
+import { Wrapper } from "./wrappers";
+import { IWriteOptions, Writer } from "./writers";
+
+/**
+ * This is a convenience interface for typesetting strings using the default
```
Comment (hellochar) [inline src/typesetter.ts]: 👍

File: src/context/canvas.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * license at https://github.com/palantir/svg-typewriter/blob/develop/LICENSE
+ */
+
+import { IAnchor, ITransform } from "../writers";
+import { ITypesetterContext } from "./index";
+
+/**
+ * Options for styling text
+ */
+export interface ICanvasFontStyle {
```
Comment (hellochar) [inline src/context/canvas.ts]: In svg this is actually controlled by the css classes right? Or would it make sense to expose a generic "font style" for both svg and canvas

File: src/context/index.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+import { IRulerFactory } from "../measurers";
+import { IPenFactory } from "../writers";
+
+export interface IPenFactoryContext<T> {
+    createPen: IPenFactory<T>;
+}
+
+export interface IRulerFactoryContext {
```
Comment (hellochar) [inline src/context/index.ts]: nit: docs these interfaces plz

File: README.md [hunk 4]
```diff
@@ -1,77 +1,108 @@
...
 };
 
-// write the text to the element.
-writer.write("Hello World!", width, height, writeOptions);
+// Write the text to the elements
+svgTypesetter.write("Hello SVG!", width, height, writeOptions);
+canvasTypesetter.write("Hello Canvas!", width, height, writeOptions);
+```
+
+### Example Shared Cache
```
Comment (hellochar) [inline README.md]: this example doesn't actually show a shared cache

File: package.json [hunk 5]
```diff
@@ -45,11 +43,11 @@
     "@types/d3": "^4.4.1",
```
Comment (hellochar) [inline package.json]: nit: can remove `@types/d3`

File: preview/preview.js [hunk 6]
```diff
@@ -1,24 +1,48 @@
...
+    };
+
+    return {
+        update,
+        text: "",
+        options: writeOptions,
+    };
+}
+
+function retinaFix(ctx) {
```
Comment (hellochar) [inline preview/preview.js]: awww omg sadness

**Comment by adidahiya** [comment]
@ashwinr for SA on NPM scope ^

**Comment by ashwinr** [comment]
+1 to @adidahiya's comment, please don't create a new `palantir` scope.

**Comment by adidahiya** [comment]
How about a `@plottable` scope? I'm thinking plottable/core, plottable/svg, plottable/canvas, plottable/typesetter, etc.

**Comment by hellochar** [comment]
what's the benefit of scopes? We really only have two libraries - typesetter and plottable - for the time being.

**Comment by adidahiya** [comment]
@hellochar No explicit benefit; it just helps us use names that aren't already taken :)

File: src/context/canvas.ts (hunk 2 above)
Comment (themadcreator) [inline src/context/canvas.ts]: in SVG it's entirely controlled by CSS.

We could add some optional common params that can be applied directly to the tag like `<text style="font: 10px arial">`

File: package.json (hunk 5 above)
Comment (themadcreator) [inline package.json]: This is still needed for dev deps since it's used in tests

File: preview/preview.js (hunk 6 above)
Comment (themadcreator) [inline preview/preview.js]: Oh this is pretty standard. We'll have to do the same when we start canvas rendering plottable.

File: README.md (hunk 4 above)
Comment (themadcreator) [inline README.md]: Sure, i'll make it explicit

**Comment by adidahiya** [comment]
> Still he just published it a few months ago so I doubt he's ready to abandon it quite yet.

meh, I think it's still worth asking. he has an email for contact. https://www.npmjs.com/~kev_nz

👍  for `@plottable/typesetter`. we'll need to get the https://www.npmjs.com/~plottable user to be able to publish to that scope (this is a restriction for the public NPM registry).

**Comment by themadcreator** [comment]
I've asked the owner of the `typesetter` package if he'd be willing to give up the name. If not, we'll try `@plottable/typesetter`

**Comment by hellochar** [comment]
Are we blocked here on that naming thing? It'd be nice to get the animation removal (since it removes clipPath which is like 50% of bp-charts' rendering time) soon

**Comment by themadcreator** [comment]
Looks like the package name "typesettable" is not taken, and it has a certain symmetry with "plottable" so i'm gonna go with that.

**Pull request merged**